dasharo-stability/capsule-update.robot

```diff
@@ -0,0 +1,96 @@
+Library             Process
+Library             String
+Library             Telnet    timeout=20 seconds    connection_timeout=120 seconds
+Library             SSHLibrary    timeout=90 seconds
+Library             RequestsLibrary
+# TODO: maybe have a single file to include if we need to include the same
+# stuff in all test cases
+Resource            ../variables.robot
+Resource            ../keywords.robot
+Resource            ../keys.robot
+
+# TODO:
+# - document which setup/teardown keywords to use and what are they doing
+# - go threough them and make sure they are doing what the name suggest (not
+# exactly the case right now)
+Suite Setup         Run Keywords
+...                     Prepare Test Suite
+...                     AND
+...                     Skip If    not ${CAPSULE_UPDATE_SUPPORT}    Capsule Update not supported
+...                     AND
+...                     Flash Firmware If Not QEMU
+Suite Teardown      Run Keywords
+...                     Flash Firmware If Not QEMU
+...                     AND
+...                     Log Out And Close Connection
+
+
+*** Test Cases ***
+CUP001.001 Capsule Update (Ubuntu)
+    [Documentation]    This test aims to verify if Capsule Update process works.
+    Power On
+    Boot System Or From Connected Disk    ubuntu
+    Login To Linux
+    Switch To Root User
+
+    Set Prompt For Terminal    root@${UBUNTU_HOSTNAME}:/home#
+    Execute Command In Terminal    cd ..
+    Set Prompt For Terminal    root@${UBUNTU_HOSTNAME}:/#
+    Execute Command In Terminal    cd ..
+    Execute Command In Terminal    rm -r capsule_testing
+    Execute Command In Terminal    mkdir capsule_testing
+    Execute Command In Terminal    chmod 777 capsule_testing
+
+    Send File To DUT    ./dasharo-stability/capsule-update-files/CapsuleApp.efi    /capsule_testing/CapsuleApp.efi
+    Send File To DUT    ${CAPSULE_FW_FILE}    /capsule_testing/dasharo.cap
+
+    Power On
+    ${boot_menu}=    Enter Boot Menu Tianocore And Return Construction
+    Enter Submenu From Snapshot    ${boot_menu}    UEFI Shell
+    Read From Terminal Until    Shell>
+
+    ${out}=    Execute UEFI Shell Command    smbiosview -t 0
+    ${original_bios_version}=    Extract BIOS Version    ${out}
+    Execute UEFI Shell Command    FS0:
+    Execute UEFI Shell Command    cd capsule_testing
+    Execute UEFI Shell Command    CapsuleApp.efi dasharo.cap
+
+    ${out}=    Read From Terminal Until    seconds.)
+
+    ${digit}=    Get Key To Press    ${out}
+    Write Bare Into Terminal    ${digit}
+
+    Set DUT Response Timeout    120s
+    ${boot_menu}=    Enter Boot Menu Tianocore And Return Construction
+    Enter Submenu From Snapshot    ${boot_menu}    UEFI Shell
+    Read From Terminal Until    Shell>
+    ${out}=    Execute UEFI Shell Command    smbiosview -t 0
+    ${updated_bios_version}=    Extract BIOS Version    ${out}
+
+    Should Not Be Equal    ${original_bios_version}    ${updated_bios_version}
```

Comment:
SMBIOS is populated by firmware based on how the firmware was built.  It doesn't know whether it was delivered in a form of a capsule or not.

Instead of using UINT_MAX, we could make an extra RC version (e.g., `rc0`) that's not even published publicly, but that could be flashed before checking any other RC and the release itself.  In terms of these tests, this would mean that `./scripts/capsules/capsule_update_tests.sh` should take 2 capsules as input to be able to create suitable test capsules.

Reply:
So what are the next steps? @SergiiDmytruk could you clarify?

Do you need to make changes in the capsule, prepare `rc0` or @JanPrusinowski just in the tests? Or both. 

Additionally, could you explain why these changes are necessary and what the benefits would be at this point?

Reply:
I’m wondering if this might be more of a feature request rather than a bug.  If you agree, could you please @SergiiDmytrukk create an issue proposing changes to the firmware code and tests? This will help clarify what needs to be done and where.

This will be a good place for a discussion  with @krystian-hebel , so there will be an optimal resolution for this aspect,  but let’s ensure this doesn’t delay the completion of the current phase.

Please link the issue here once it’s created so we can resolve this thread and address this part in the next phase.

Reply:
I like the idea of `rc0`, although there are few important aspects:
- `rc0` would have to be build for each RC from the current code at that point, to properly test potential changes in capsule handling code between RCs
- if those would be stored, just having it named `*-rc0` may not be enough - with each new RC a new `rc0` would overwrite the previous one, this could hide some problems or make it impossible to reproduce the previous results.

Few future improvement ideas (not part of this task) could include:
- building the images as part of regression tests
  - possible reproducibility issues
  - signing not feasible
- downloading the binaries automatically from cloud as preparation to tests, to reduce the number of images that have to be passed when starting tests
  - good for releases, but not for dev-tests

Reply:
I am not sure what you mean. Anyway I do not know of any other way to check if the FW has actually changed... 
Qemu:
![image](https://github.com/user-attachments/assets/5aa63f4c-a75a-4522-99c1-5bb5bdc60779)
MSI:
![image](https://github.com/user-attachments/assets/289242cf-84cb-41a8-846d-45512e107c5c)
This method doesn't yield any results.

For now i have manually created FW files with _capsule_testing added to version and increased version numbers.

Reply:
Tests need 2 firmware builds with different versions which @JanPrusinowski is already doing.  So there is no need to even update that script because "initial firmware" is flashed separately.

The rest is about release policy and testing releases.  I think no matter whichever way it goes, tests in current form should work.

> * `rc0` would have to be build for each RC from the current code at that point, to properly test potential changes in capsule handling code between RCs

Wouldn't testing `rc0 -> rc1` and then `rc1 -> rc2` work and be enough?

Reply:
> Wouldn't testing `rc0 -> rc1` and then `rc1 -> rc2` work and be enough?

In most cases yes, but it wouldn't work if it was `rc1` that actually introduced an issue.

On the other hand, to do what users will do, we would have to test it with "initial firmware" being previous release, and preferably also lowest supported version as specified in JSON. However, as we don't have any previous releases with capsule support yet, we can add it later.

> Anyway I do not know of any other way to check if the FW has actually changed...

`smbiosview -t 0` (field `BIOS Version:`)

Reply:
@krystian-hebel could you specify the exact changes that Jan needs to implement in this PR?

If no changes are required from this discussion, let’s move it to a new issue and close this thread. Let’s avoid complicating this task further.

Reply:
@BeataZdunczyk, created https://github.com/Dasharo/open-source-firmware-validation/issues/494, resolving this as there is nothing to do for now.